Feature/quadratic problem scale

michalk8 opened the pull request:
In this PR:
- add `scale_cost` to `QuadraticProblem` which updates the `scale_cost` in all geometries
  when passed as `False` (default), it keeps the orignal `scale_cost` in the geometries
- allow `scale_cost` in geometries to be `bool`: if `True` a default is used (`mean` for `LRGeometry` and `PointCloud`, `max_cost` for `Geometry`, but am not very opinionated here)
- update CI to run on 3.8 (numpy support for 3.7 ended last December) and use only 2 workers (win/linux machines have 2 CPUs only, see [here](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners))
- fix some warnings in tests
- increase tolerance in `test_apply_squared` from `1e-4` to `5e-4` (was locally failing, also on `main`)

marcocuturi commented on ott/core/quad_problems.py:
```diff
@@ -106,6 +107,10 @@ def __init__(self,
+      scale_cost
+        option to rescale the cost matrices. If `True`, use the default for
+        each geometry. If `False`, keep the original scaling in geometries.
+        If `None`, do not scale the cost matrices.
```
mention when 'str' is passed

michalk8 replied:
Done.

marcocuturi commented on ott/core/quad_problems.py:
```diff
@@ -493,6 +498,17 @@ def update_epsilon_unbalanced(epsilon, transport_mass):
   return updated_epsilon
 
 
+def update_geom_scale_cost(geom: Optional[geometry.Geometry],
```
maybe this should be in geometry.py and take self as first argument?

michalk8 replied:
Originally, I was against this (as the method either returns unmodified `self` or a new instance), but I changed my mind and think your suggestion is better.
Done as a private method `_set_scale_cost(self, scale_cost)`.

marcocuturi commented on ott/geometry/geometry.py:
```diff
@@ -73,14 +73,15 @@ def __init__(self,
     self._relative_epsilon = relative_epsilon
     self._scale_epsilon = scale_epsilon
-    self._scale_cost = scale_cost
+    self._scale_cost = "max_cost" if scale_cost is True else scale_cost
```
would rather use mean for consistency.

michalk8 replied:
Done.

marcocuturi commented on tests/core/gromov_wasserstein_test.py:
```diff
@@ -231,6 +233,25 @@ def test_gw_lr_fused(self):
+    geom_y_scaled = pointcloud.PointCloud(self.y, scale_cost=scale_cost)
+    geom_xy_scaled = pointcloud.PointCloud(self.xx, self.yy, scale_cost=scale_cost)
+
+    gt = gromov_wasserstein.gromov_wasserstein(geom_xx=geom_x_scaled, geom_yy=geom_y_scaled, geom_xy=geom_xy_scaled,
```
80 chars (here and elsewhere)

michalk8 replied:
Done, although in #43 I will argue to go for `120`.

marcocuturi commented on tests/tools/sinkhorn_divergence_test.py:
```diff
@@ -187,8 +187,8 @@ def test_segment_sinkhorn_result(self, shuffle):
       # still maintain the segment assignments:
-      idx_x = jax.random.shuffle(rngs[2], jnp.arange(x.shape[0] * 2))
-      idx_y = jax.random.shuffle(rngs[3], jnp.arange(y.shape[0] * 2))
+      idx_x = jax.random.permutation(rngs[2], jnp.arange(x.shape[0] * 2), independent=True)
```
thanks!

marcocuturi reviewed (commented):
Thanks Michal!

michalk8 commented:
When updating the docstrings for `scale_cost`, I noticed there are many warnings with the docs (many lists/equations were incorrectly rendered), so I fixed it as well. Some warnings are still left, though they are not related to syntax.
Furthermore, some link to classes were broken (e.g., Linear/QuadradProblem), fixed whatever I could. Biggest improvement can be seen in the docs for `sinkhorn` function.

I also noticed that people can't click on, the functions from the classes, e.g. [PointCloud](https://ott-jax.readthedocs.io/en/latest/_autosummary/ott.geometry.pointcloud.PointCloud.html#ott.geometry.pointcloud.PointCloud).
I've added a very simple template which fixes it. Overall, docs will need a proper fix, this is just a triage (e.g. the package versions differ in multiple places).

Lastly, some tests are still failing, but they are afaik unrelated to this PR. I think the tolerances are too low, will check later (locally all pass).

marcocuturi commented:
I think the tests are a bit flaky indeed. On github this seems to be a timeout issue.

Thanks a ton for this PR Michal, this is really great!
